### ISSUE TITLE
KafkaConsumer: support per-partition queue consumption

### ISSUE THREAD

**Issue by FDePourcq** [edenhill/librdkafka#539]
Hi, question:
In a multithreaded application that listens to 1 kafka topic I have the choice between 
1 consumer for each thread versus 1 consumer for all threads.

But is the first scenario, 1 consumer for each thread, even possible? I see a lot of static keywords in the code. (And that global opaque value per singleton topic doesn't motivate either. The amount of connections this would open in the background could make it entirely foolish.)

**Comment by edenhill** [comment]
Do you want each app thread to consume a single (set of) partition(s)?
Or do you want to consume all partitions from a single thread?

You are right in that each consumer instance will create 1 + N threads where N is the number of brokers.

**Comment by FDePourcq** [comment]
each partition would be assigned to just one consumer, there are enough partitions... so each app thread would be the sole consumer of the partitions assigned to it

**Comment by edenhill** [comment]
Okay, are you planning on using the high-level consumer or the simple consumer?
The former puts all messages in the same queue so that isn't perfect for your use-case, but the latter allows consumption per-partition and thus per-thread.

**Comment by FDePourcq** [comment]
If I understand it well ... the high-level kafkaconsumer (both c++ kafkaconsumer as C-kafkacat-kafkaconsumer) have a singleton queue ... which would make it impossible to have one for each thread ... but the app could internally still organize itself to distribute the messages after arrival among the threads.

And alternatively I could go to the old style simple consumer, 1 per thread, but then I would have to do the rebalancing and offset-management from scratch. (and the whole offset-commit-to-broker-story would be entirely gone too)

To me it looks easiest to have 1 high-level-kafkaconsumer and distribute the messages behind it to the different threads. (Hoping and suspecting that it would be no performance-bottleneck.)

**Comment by FDePourcq** [comment]
( the code i focus on is based on the highlevel kafkaconsumer-code in kafkacat.c )

**Comment by edenhill** [comment]
KafkaConsumer, right, but not kafkacat - kafkacat simply enumerates the available partitions and consumes all of them , but there is no consumer group balancing involved (yet!).

If you need rebalancing you really dont want to do that yourself, then it is better to use KafkaConsumer as-is and dispatch messages to your worker threads from a single consumer->consume() thread precisely as you say.

Having said that it would not be a huge change to allow per-partition consumption in the high-level consumer, or more precisely, partition queue message routing.

**Comment by edenhill** [comment]
Something like:
````
consumer->set_partition_queue("topic", 0, someQueue);
.....

in worker thread:
while (run) {
   msg =  someQueue->consume();
 ..
}
````

**Comment by FDePourcq** [comment]
ah you are thinking of incorporating the dispatching to different threads that way? (with still a singleton queue centrally) (this would move a bunch of messagepassing trouble between threads to your domain )

**Comment by FDePourcq** [comment]
Some thoughts; I figured that, with one master-highlevel-kafkaconsumer dispatching messages to multiple slave threads:
- this is another place where there can be a partition-to-slavethread-assignment with rebalancing and rebalancing strategies :)
- it is easy to get a situation wherein the singleton consumer will block on the slowest slavethread.
- per thread are 2 queues needed: the other one in the other direction to advance the offset.

**Comment by dwieland** [comment]
+1 on this.

I use the HL-Consumer for rebalancing and dispatch each partition to its own thread.

> it is easy to get a situation wherein the singleton consumer will block on the slowest slavethread.

This is the real problem with this solution. Without another big queue for each thread a series of slow tasks on one partition can really slow down the whole processing.

I did some quick changes where I hook into rd_kafka_assign inside my rebalance_cb and forward the partition fetch queues to my own rd_kafka_queue_t's instead of the client groups queue. This seems to work quite well, but I don't know what other problems I might produce with that as the op queues are still forwarded to the client group.

**Comment by edenhill** [comment]
addressed by #953

**Comment by edenhill** [comment]
Fixed on master
